packages/pds/src/migrate-script/index.ts

```diff
@@ -0,0 +1,417 @@
+
+  console.log(`transfer (${status.did}, ${status.phase})`)
+  if (status.phase < TransferPhase.transferred) {
+    const importedRev = await lockAndTransfer(ctx, db, pds, status)
+    status.importedRev = importedRev
+    status.phase = TransferPhase.transferred
+    await updateStatus(db, status)
+  }
+
+  console.log(`prefs (${status.did}, ${status.phase})`)
+  if (status.phase < TransferPhase.preferences) {
+    try {
+      await transferPreferences(ctx, pds, status.did)
+    } catch (err) {
+      await db
+        .insertInto('failed_pref')
+        .values({ did: status.did })
+        .onConflict((oc) => oc.doNothing())
+        .execute()
+    } finally {
+      status.phase = TransferPhase.preferences
+      await updateStatus(db, status)
+    }
+  }
+
+  console.log(`takedowns (${status.did}, ${status.phase})`)
+  if (status.phase < TransferPhase.takedowns) {
+    await transferTakedowns(ctx, db, pds, status.did, adminToken)
+    status.phase = TransferPhase.completed
+    await updateStatus(db, status)
+  }
+}
+
+const updateStatus = async (db: MigrateDb, status: Status) => {
+  return db
+    .updateTable('status')
+    .set({ ...status })
+    .where('did', '=', status.did)
+    .execute()
+}
+
+const reserveSigningKey = async (
+  pds: PdsInfo,
+  did: string,
+): Promise<string> => {
+  const signingKeyRes =
+    await pds.agent.api.com.atproto.server.reserveSigningKey({ did })
+  return signingKeyRes.data.signingKey
+}
+
+const doImport = async (
+  ctx: AppContext,
+  db: MigrateDb,
+  pds: PdsInfo,
+  did: string,
+  since?: string,
+) => {
+  const storage = new SqlRepoStorage(ctx.db, did)
+  const root = await storage.getRootDetailed()
+  if (!root) {
+    throw new Error(`repo not found: ${did}`)
+  }
+  if (since && root.rev === since) {
+    return
+  }
+  const carStream = await storage.getCarStream(since)
+
+  const importRes = await axios.post(
+    `${pds.url}/xrpc/com.atproto.temp.importRepo`,
+    carStream,
+    {
+      params: { did },
+      headers: { 'content-type': 'application/vnd.ipld.car' },
+      decompress: true,
+      responseType: 'stream',
+    },
+  )
+
+  for await (const log of importRes.data) {
+    if (typeof log === 'string') {
+      const lines = log.split('\n')
+      for (const line of lines) {
+        if (line.indexOf('failed to import blob') > 0) {
+          const cid = line.split(':')[1].trim()
+          await logFailedBlob(db, did, cid)
+        }
+      }
+    }
+    console.log(`import update for ${did}: `, log.toString())
+  }
+  return root.rev
+}
+
+const lockAndTransfer = async (
+  ctx: AppContext,
+  db: MigrateDb,
+  pds: PdsInfo,
+  status: Status,
+) => {
+  const defer = createDeferrable()
+  ctx.db
+    .transaction(async (dbTxn) => {
+      const storage = new SqlRepoStorage(dbTxn, status.did)
+      await storage.lockRepo()
+      await defer.complete
+    })
+    .catch((err) => {
+      console.error(`error in repo lock tx for did: ${status.did}`, err)
+    })
```

Comment:
If this does fail there may be some important moments below where we don't want to continue.  Could we set a variable in the `catch()` and then assert that it's not set whenever we want below?

Reply:
yup yup good looks 👍